### PR TITLE
CouchDB config initialisation and secret extraction, expose users db to components

### PR DESCRIPTION
**Pull request by boennemann**

### CODE REVIEW

File: lib/couchdb.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    })
+  }
+
+  var couch = request.defaults({
+    baseUrl: config.db.url,
+    json: true
+  })
+
+  async.waterfall([
+    function (callback) {
```
Comment (janl) [inline lib/couchdb.js]: style nit: wanna name these anon functions? And maybe extract from `async.waterfall()` so it is:

```js
var foo = function(callback) { … }
var bar = function(callback) { … }

async.waterfall([
  foo,
  bar
], …)
```

File: lib/couchdb.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  if (!config.db.url) {
+    var storePath = join(config.paths.data, 'config.json')
+    var storeExists = existsSync(storePath)
+    var store = storeExists && jsonfile.readFileSync(storePath, {
+      throws: false
+    }) || {}
+    var secret = store.couch_httpd_auth_secret
+
+    if (!secret) {
+      secret = randomstring.generate({
```
Comment (janl) [inline lib/couchdb.js]: In the old hoodie-server we had it that at this point, we already had set up an admin user, and the secret created.

What is the plan for setting up an admin user in the new hoodie-server?

Comment (boennemann) [inline lib/couchdb.js]: Currently we're expecting the admin user to be created and to be passed via the url. `!config.db.url` means that it's using pouchdb and that's why I'm generating the secret.

File: lib/couchdb.js (hunk 1 above)
Comment (boennemann) [inline lib/couchdb.js]: That's exactly what I'm about to do, also because that makes it more testable. Just wanted to get this up first :)

**Comment by boennemann** [comment]
@janl @gr2m @christophwitzko Please review.

File: lib/config.js [hunk 3]
```diff
@@ -46,12 +46,13 @@ module.exports = function (options) {
...
     }
+  } else {
+    log.warn('config', 'No database URL provided, falling back to PouchDB')
   }
 
   if (options.inMemory) {
     config.db.db = require('memdown')
   }
 
-  require('deep-freeze')(config)
```
Comment (gr2m) [inline lib/config.js]: just wondering, why did you remove that?

**Comment by gr2m** [comment]
LGTM

File: lib/config.js (hunk 3 above)
Comment (boennemann) [inline lib/config.js]: I introduced this mostly for debugging, so I could find all spots where it was initially modified all over the place, also inside the plugins api stuff.
Most of this isn't in hoodie-server anymore and now I deleted it because I needed to modify the config object with the server's secret.
As everything is now encapsulated in components it might no longer be necessary, but we can add it again later if necessary.

File: lib/hapi.js [hunk 4]
```diff
@@ -19,13 +28,10 @@ module.exports = function (config) {
       options: {config: config}
     }
   })/*, [{
-    register: require('hoodie-server-store'),
-    options: _.defaults({prefix: '/hoodie/store'}, defaultOpts)
-  }, {
-    register: require('hoodie-server-account/routes/session'),
+    register: require('@hoodie/server-account-node-sessions'),
```
Comment (gr2m) [inline lib/hapi.js]: It’s out of scope for this PR, but the new options required for `hoodie-server-account` are `options.db` and `options.admins` as described here: https://github.com/hoodiehq/hoodie-server-account/tree/node-sessions/plugin#optionsdb

Does that make sense from your perspective? We use the `options.admins` map to see if someone tries to sign in as an admin user, as admins don’t necessarly have a `_users` doc

**Comment by janl** [comment]
+1

File: lib/hapi.js (hunk 4 above)
Comment (boennemann) [inline lib/hapi.js]: The current options look like this: `https://github.com/hoodiehq/hoodie-server/blob/users-db/lib/hapi.js#L9-L14`
Your link is missing the PouchDB factory, and the users db is called users, not just generic `db`.
Once we've agreed on this we can easily add the admins map in another PR then I guess.

**Comment by boennemann** [comment]
Okay so can we get this one out, and add the admins thing in another PR?
Whoever feels like it may merge then :)

**Comment by gr2m** [comment]
yes, ship it :)
